Add logging for all devices

mundruid opened the pull request:
- Add log utility to print error messages when exceptions occur and other information messages
- Log files are created
- Addresses issue https://github.com/networktocode/pyntc/issues/110

jvanderaa commented:
@mundruid  This may need to get changed in the PR to using develop instead of main to help with the diffs

mundruid commented:
> @mundruid This may need to get changed in the PR to using develop instead of main to help with the diffs

Thanks, @jvanderaa . I keep forgetting that the default PR in this repo is main and not develop. Changed.

itdependsnetworks commented:
I am not sure if there is something unique about this logging (I see you are importing from pyntc.log as an example), but generally you are meant to use `%s` style formatting and not `f-string` formatting for logging.

mundruid commented:
> I am not sure if there is something unique about this logging (I see you are importing from pyntc.log as an example), but generally you are meant to use `%s` style formatting and not `f-string` formatting for logging.

@itdependsnetworks I am happy to make the change. I was wondering why we would prefer `%` instead of `f-string`?

itdependsnetworks commented:
It has to do with the lack of lazy loading in f-strings. This does a better job of explaining https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

mundruid commented:
> It has to do with the lack of lazy loading in f-strings. This does a better job of explaining https://medium.com/swlh/why-it-matters-how-you-log-in-python-1a1085851205

@itdependsnetworks I have addressed the comment and replaced all the f-strings. Sorry it took so long. Let me know what you think.

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -535,12 +582,20 @@ def config(self, command, **netmiko_args):
                 command_responses.append(command_response)
                 self._check_command_output_for_errors(cmd, command_response)
         except TypeError as err:
+            log.error("Host %s: Netmiko Driver's {err.args[0]}", self.host, err.args[0])
```
```suggestion
            log.error("Host %s: Netmiko Driver's %s", self.host, err.args[0])
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -535,12 +582,20 @@ def config(self, command, **netmiko_args):
         except CommandError as err:
             if not original_command_is_str:
+                log.error(
+                    "Host %s: Commands {entered_commands} returned the error {err.cli_error_msg}",
```
```suggestion
                    "Host %s: Commands %s returned the error %s",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -535,12 +582,20 @@ def config(self, command, **netmiko_args):
+                )
                 raise CommandListError(entered_commands, cmd, err.cli_error_msg)
             else:
+                log.error("Host %s: Commands {entered_commands} returned the error %s", self.host, entered_commands)
```
```suggestion
                log.error("Host %s: Commands %s returned the error %s", self.host, entered_commands, err.cli_error_msg)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -552,6 +607,12 @@ def config(self, command, **netmiko_args):
             return command_responses[0]
 
+        log.info(
+            "Host %s: List of config commands %s received responses {command_response}.",
```
```suggestion
            "Host %s: List of config commands %s received responses %s.",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -580,7 +641,7 @@ def config_list(self, commands, **netmiko_args):  # noqa: D401
             >>>
         """
-        warnings.warn("config_list() is deprecated; use config.", DeprecationWarning)
+        log.warning("config_list() is deprecated; use config.", DeprecationWarning)
```
```suggestion
        log.warning("config_list() is deprecated; use config.")
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -626,10 +693,12 @@ def connected(self):
         Returns:
             bool: True if the device is connected, else False.
         """
+        log.debug("Host %s: Connection status {self._connected}.", self.host, self._connected)
```
```suggestion
        log.debug("Host %s: Connection status %s.", self.host, self._connected)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -626,10 +693,12 @@ def connected(self):
 
     @connected.setter
     def connected(self, value):
+        log.debug("Host %s: Device connected is %s.", self.host, value)
```
```suggestion
        log.debug("Host %s: Device connection status is %s.", self.host, value)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -673,8 +742,16 @@ def disable_wlans(self, wlan_ids):
             if not wlans_to_validate.issubset(post_disabled_wlans):
                 desired_wlans = wlans_to_validate.union(disabled_wlans)
+                log.error(
+                    "Host %s: WLANS not disabled, with desired WLANs {desired_wlans} and post disabled WLANs {post_disabled_wlans}",
```
```suggestion
                    "Host %s: Unable to disable WLANS, with desired WLANs %s and post disabled WLANs %s",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -673,8 +742,16 @@ def disable_wlans(self, wlan_ids):
+                )
                 raise WLANDisableError(self.host, desired_wlans, post_disabled_wlans)
 
+        log.info("Host %s: All WLANs with IDs {disabled_wlans} were disabled.", self.host, disabled_wlans)
```
```suggestion
        log.info("Host %s: All WLANs with IDs %s were disabled.", self.host, disabled_wlans)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -699,6 +776,7 @@ def disabled_wlans(self):  # noqa: D403
             >>>
         """
         disabled_wlans = [wlan_id for wlan_id, wlan_data in self.wlans.items() if wlan_data["status"] == "Disabled"]
+        log.info("Host %s: Disabled WLAN IDs: {disabled_wlans}", self.host, disabled_wlans)
```
```suggestion
        log.info("Host %s: Disabled WLAN IDs: %s", self.host, disabled_wlans)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -757,6 +837,13 @@ def enable_wlans(self, wlan_ids):
                 desired_wlans = wlans_to_validate.union(enabled_wlans)
+                log.error(
+                    "Host %s: WLANS not enabled,\n"
+                    f"with desired WLANs {desired_wlans} and post enabled WLANs {post_enabled_wlans}",
```
```suggestion
                    "with desired WLANs %s and post enabled WLANs %s",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -757,6 +837,13 @@ def enable_wlans(self, wlan_ids):
             if not wlans_to_validate.issubset(post_enabled_wlans):
                 desired_wlans = wlans_to_validate.union(enabled_wlans)
+                log.error(
+                    "Host %s: WLANS not enabled,\n"
```
```suggestion
                    "Host %s: Unable to enable WLANs,\n"
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -783,6 +870,7 @@ def enabled_wlans(self):  # noqa: D403
             >>>
         """
         enabled_wlans = [wlan_id for wlan_id, wlan_data in self.wlans.items() if wlan_data["status"] == "Enabled"]
+        log.info("Host %s: List of enabled WLAN IDs: {enabled_wlans}", self.host, enabled_wlans)
```
```suggestion
        log.info("Host %s: List of enabled WLAN IDs: %s", self.host, enabled_wlans)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -859,20 +947,35 @@ def file_copy(
             )
         except CommandListError as error:
+            log.error(
+                "Host %s: File transfer error {FileTransferError.default_message}\n\n{error.message}",
```
```suggestion
                "Host %s: File transfer error %s\n\n%s",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -859,20 +947,35 @@ def file_copy(
                 response = self.show("y", auto_find_prompt=False, delay_factor=delay_factor)
         except CommandError as error:
+            log.error(
+                "Host %s: File transfer error {FileTransferError.default_message}\n\n{error.message}",
```
```suggestion
                "Host %s: File transfer error %s\n\n%s",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -859,20 +947,35 @@ def file_copy(
 
         if "File transfer is successful" not in response:
-            raise FileTransferError(message=f"Did not find expected success message in response, found:\n{response}")
+            log.error("Host %s: Did not find expected success message in response, found:\n%s", self.host, response)
```
```suggestion
            log.error("Host %s: Did not find expected success message in file transfer response, found:\n%s", self.host, response)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -938,14 +1041,25 @@ def install_os(self, image_name, controller="both", save_config=True, disable_wl
+                    hostname=f"Host {self.host}: {self.host}-standby", desired_boot=f"{image_name}-{peer_redundancy}"
+                )
 
+            log.info("Host %s: OS image %s installed successfully.", self.host, image_name)
```
```suggestion
            log.info("Host %s: OS image %s installed.", self.host, image_name)
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1105,6 +1227,7 @@ def redundancy_mode(self):
         """
         ha = self.show("show redundancy summary")
         ha_mode = re.search(r"^\s*Redundancy\s+Mode\s*=\s*(.+?)\s*$", ha, re.M)
+        log.debug("Host %s: Redundancy mode: {ha_mode.group(1).lower()}", self.host, ha_mode.group(1).lower())
```
```suggestion
        log.debug("Host %s: Redundancy mode: %s", self.host, ha_mode.group(1).lower())
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
                 command_responses.append(command_response)
                 self._check_command_output_for_errors(cmd, command_response)
         except TypeError as err:
+            log.error("Host %s: Netmiko Driver's {err.args[0]}", self.host, err.args[0])
```
```suggestion
            log.error("Host %s: Netmiko Driver's %s", self.host, err.args[0])
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
         except CommandError as err:
             if not original_command_is_str:
+                log.error(
+                    "Host %s: Command error for commands {entered_commands} with message {err.cli_error_msg}.",
```
```suggestion
                    "Host %s: Command error for commands %s with message %s.",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
                 raise CommandListError(entered_commands, cmd, err.cli_error_msg)
             else:
+                log.error(
+                    "Host %s: Command error for commands {entered_commands} with message %s.",
```
```suggestion
                    "Host %s: Command error for commands %s with message %s.",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1256,18 +1385,36 @@ def show(self, command, expect_string=None, **netmiko_args):
             return command_responses[0]
 
+        log.debug(
+            "Host %s: Successfully executed command 'show' with responses {command_responses}.",
```
```suggestion
            "Host %s: Successfully executed show commands with responses %s.",
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1382,8 +1530,10 @@ def transfer_image_to_ap(self, image, timeout=None):
         if not self._ap_images_match_expected("primary", image):
-            raise FileTransferError(f"Unable to set all APs to use {image}")
+            log.error("Host %s: Unable to set all APs to use %s", self.host, image)
+            raise FileTransferError
```
```suggestion
            raise FileTransferError(f"Unable to set all APs to use {image}")
```

jmcgill298 commented on pyntc/devices/aireos_device.py:
```diff
@@ -1404,6 +1554,7 @@ def uptime(self):
         hours += days * 24
         minutes += hours * 60
         seconds = minutes * 60
+        log.debug("Host %s: Device has been up for %s seconds", self.host, seconds)
```
```suggestion
        log.debug("Host %s: Device has been up for %d seconds", self.host, seconds)
```

jmcgill298 reviewed (changes requested):
@mundruid I started with these, so far I have only made it through the aireos driver

mundruid commented:
> @mundruid I started with these, so far I have only made it through the aireos driver

@jmcgill298 Thanks for the review!

jeffkala commented:
@jmcgill298 @mundruid I know its a longshot being this PR is so old, but was this ready to go? Worth revisiting to get it over the finish line or just close?

mundruid commented:
@jeffkala On my part, this was ready after addressing Ken's comment about the not using `f-strings` in all the logs. I have added the tests and all the logging. However, It is a long one to review and I think that is why we have not merged it yet.

jmcgill298 commented:
@jeffkala I put several comments above where things do not look quite right. I did not get through all of the updates, but I suspect similar comments to be repeated

jeffkala commented:
@mundruid @jmcgill298 thanks, will try and work on getting the rest updated and get this PR over the finish line.

jeffkala commented:
taking it over the finish line in this pr #271
